rholang/examples/linking/v0.2/packages/LinkedList.rho

```diff
@@ -0,0 +1,149 @@
+    }
+  } |
+  contract @{*LinkedList | "map"}(@list, function, return) = {
+    new combinator in {
+      contract combinator(@head, @accumulatedValue, return) = {
+        new mappedHead in {
+          function!(head, *mappedHead) |
+          for(@x <- mappedHead) {
+            @{*LinkedList | "prepend"}!(x, accumulatedValue, *return)
+          }
+        }
+      } | new reversedMappedList in {
+        @{*LinkedList | "fold"}!(list, [], *combinator, *reversedMappedList) |
+        for (@rml <- reversedMappedList) {
+          @{*LinkedList | "reverse"}!(rml, *return)
+        }
+      }
+    }
+  } |
+  contract @{*LinkedList | "foreach"}(@list, proc, isDone) = {
+    new combinator, nilReturn in {
+      contract combinator(@head, @accumulatedValue, return) = {
+        //need return flag from `proc` in order to guarantee execution order
+        new isDone in { proc!(head, *isDone) | for(_ <- isDone){ return!(Nil) } }
+      } | 
+      @{*LinkedList | "fold"}!(list, Nil, *combinator, *nilReturn) | for(_ <- nilReturn){ isDone!(true) }
+    }
+  } |
+  contract @{*LinkedList | "get"}(@list, @index, return) = {
+    new combinatorAndCondition in {
+      contract combinatorAndCondition(@head, @accumulatedValue, return) = {
+        match accumulatedValue {
+          [i, _] => { return!([i == index, [i + 1, head]]) }
+        }
+      } |
+      new result in {
+        @{*LinkedList | "partialFold"}!(list, [0, Nil], *combinatorAndCondition, *result) |
+        for(@r <- result) {
+          match r {
+            [true, [_, v]] => { return!(v) }
+            _ => { return!(Nil) } //index out of bounds
+          }
+        }
+      }
+    }
+  } |
+  contract @{*LinkedList | "indexOf"}(@list, @value, return) = {
+    new combinatorAndCondition in {
+      contract combinatorAndCondition(@head, @accumulatedValue, return) = {
+        return!([head == value, accumulatedValue + 1])
+      } |
+      new result in {
+        @{*LinkedList | "partialFold"}!(list, -1, *combinatorAndCondition, *result) |
+        for(@r <- result) {
+          match r {
+            [true, i] => { return!(i) }
+            _ => { return!(-1) } //element not found
+          }
+        }
+      }
+    }
+  } |
+  contract @{*LinkedList | "length"}(@list, return) = {
+    new combinator in {
+      contract combinator(@head, @accumulatedValue, return) = {
+        return!(accumulatedValue + 1)
+      } | 
+      @{*LinkedList | "fold"}!(list, 0, *combinator, *return)
+    }
+  } |
+  contract @{*LinkedList | "range"}(@start, @end, return) = {
+    new loop in {
+      contract loop(@i, @list) = {
+        match i < start {
+          true => { return!(list) } 
+          _ => {
+            loop!(i - 1, [i, list])
+          }
+        }
+      } | loop!(end - 1, [])
+    }
+  } |
+  import RhoClass in {
+    contract LinkedList(@method, @args) = {
+      @{*RhoClass | "apply"}!({*LinkedList | method}, args)
+    }
+  }
+}
```

Comment:
Can you arrange to put newlines at the end of your files? It's kind of annoying.